COMPACT_DENYLIST_FORMAT.md

```diff
@@ -0,0 +1,457 @@
+
+# Block some subpaths with exceptions: last-matching-rule wins (!)
+/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked*
+!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blockednot
+!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/not
+!/ipfs/QmUboz9UsQBDeS6Tug1U8jgoFkgYxyYood9NDyVURAY9pK/blocked/exceptions*
+
+# Block IPNS domain name
+/ipns/domain.example
+
+# Block IPNS domain name and path
+/ipns/domain2.example/path
+
+# Block IPNS key - blocks wrapped multihash.
+/ipns/k51qzi5uqu5dhmzyv3zac033i7rl9hkgczxyl81lwoukda2htteop7d3x0y1mf
+
+# Block all mime types with exceptions
+/mime/image/*
+!/mime/image/gif
+
+# Legacy CID double-hash block
+# sha256(bafybeiefwqslmf6zyyrxodaxx4vwqircuxpza5ri45ws3y5a62ypxti42e/)
+# blocks only this CID
+//d9d295bde21f422d471a90f2a37ec53049fdf3e5fa3ee2e8f20e10003da429e7
+
+# Legacy Path double-hash block
+# Blocks bafybeiefwqslmf6zyyrxodaxx4vwqircuxpza5ri45ws3y5a62ypxti42e/path
+# but not any other paths.
+//3f8b9febd851873b3774b937cce126910699ceac56e72e64b866f8e258d09572
+
+# Double hash CID block
+# base58btc-sha256-multihash(QmVTF1yEejXd9iMgoRTFDxBv7HAz9kuZcQNBzHrceuK9HR)
+# Blocks bafybeidjwik6im54nrpfg7osdvmx7zojl5oaxqel5cmsz46iuelwf5acja
+# and QmVTF1yEejXd9iMgoRTFDxBv7HAz9kuZcQNBzHrceuK9HR etc. by multihash
+//QmX9dhRcQcKUw3Ws8485T5a9dtjrSCQaUAHnG4iK9i4ceM
+
+# Double hash Path block using blake3 hashing
+# base58btc-blake3-multihash(gW7Nhu4HrfDtphEivm3Z9NNE7gpdh5Tga8g6JNZc1S8E47/path)
+# Blocks /ipfs/bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a/path
+# /ipfs/bafyb4ieqht3b2rssdmc7sjv2cy2gfdilxkfh7623nvndziyqnawkmo266a/path
+# /ipfs/f01701e20903cf61d46521b05f926ba1634628d0bba8a7ffb5b6d5a3ca310682ca63b5ef0/path etc...
+# But not /path2
+//QmbK7LDv5NNBvYQzNfm2eED17SNLt1yNMapcUhSuNLgkqz
+```
+
+#### High level list format
+
+A denylist is made of an optional header and a list of blockitems separated by newlines. Comment lines start with `#`. Empty lines are allowed.
+
+```
+<header>
+---
+<block_item1> [hint_list]
+
+# comment
+<block_item2> [hint_list]
+...
+```
+
+#### Header
+
+The list header is a YAML block:
+
+- Must be valid YAML
+- Fully optional
+- 1KB maximum size
+- Delimited by a line containing `---` at the end (document separator)
+
+Known-fields (they must be lowercase):
+
+- `version`: the denylist format version. Defaults to 1 when not specified.
+- `name`
+- `description`
+- `author`
+- `hints`: a map of *hints*. See section below for known hints
+
+#### Hints
+
+A *hint* is a key-value duple associated to the denylist as a whole (part of the header), or to a specific \<block_item\>.
+
+Header hints can be used to set denylist-wide options or information that
+implementations can choose to interpret or not.
+
+#### List body
+
+A denylist is made of lines which are made by a *block items* followed by zero or more space-separated hints.
+
+Lines should not be longer than 2MiB including the "\n" delimiter.
```

Comment:
Is the expectation that the implementation will bail on longer lists? Or ignore after a cut-off?

Reply:
It's a limit per line. There is an issue here: you can always have valid IPLD paths that are longer than whatever the limit it, so those would be unblockable. I'm not sure what to make of that. On the other side, having a 50GB line might be a way to exhaust memory. 2MiB seemed like a compromise, but could be higher and be fine too I guess.

Reply:
Wait, does this mean that the *block items*  (referents of the various kinds of entries) need to be under 2MiB, or that lines with 1.9MiB of annotations are valid?

Reply:
The line of text should be less than 2MiB. How that is divided between block_item and annotations is not specified. IMHO, it is a bit easier for a parser implementation to enforce limits on the basic unit of a line, rather than having separate limits for the block_items or allocations, even if that would perhaps be more "correct".